### ISSUE TITLE
unified vs tasks - summary of my view

### ISSUE THREAD

**Issue by dobkeratops** [bbernhard/imagemonkey-core#242]
ok I wanted to make a dedicated thread for this - sticking to constructive criticism and suggestions on how you might be able to steer the project.

**Now you've got the unified mode, it exposes the shortcomings of the task idea**

my recommendations that you (i) focus on the unified modewind down the task idea and consider it an evolutionary step to get to where you are, and , and (ii) find other ways for casual users to contribute

What's happened is you've had to evolve from the original idea - *just one label per image* - to multiple labels. The unified mode is what you really need for multiple labels.

You appear to have a concept of a **Casual vs a Pro user**: my view is that anything to do with *polygonal outlines* requires what you currently call a "Pro user".  

I think The mistake here is thinking that tasks allow 'Casual' to work with polygons. In reality they cripple the polygonal task. The unified mode lets you see how things fit together : you eyes can scan the area gradually - you work across the image finding and covering things.

But what is pro really? 

Labelling is like **painting**: the unified mode gives you something like layer and palette management. Tasks is like asking someone to paint but only allowing them to use each color once ("you must draw ALL the light blue in this image, then forget the image, move on." `you can label using a paint program, but you're limited to clear color coding, or label management which can get a bit insane buried in fully general paint UI.

Anyone who can use a paint program should be able to handle the unified mode? (paint doesn't have to mean expert concept artists, plenty of people can draw simple cartoons, color coded diagrams etc)

consider the basic idea of "annotate all Man, annotate all Woman" for a group of people - as 2 tasks, one user has to scan each person looking for 'Man', another looking for 'Woman' . that's 2 passes. Now consider traffic, where you have car,van,truck,bus,.. now it's 4 passes.

by contrast in the unified mode you can flick through with your eyes locked on one part to check if you had it ... and it'll be even better when you have an option to toggle show all the outlines, maybe with color coding)

**but also consider the "add labels" screen**. 

I'd view this as a stop-gap whilst the tool was evolving from the "one label per image" format.  **It s actually really difficult to verify you have all the labels without seeing the annotations** .. its like you have to keep the label list *in your head* whilst you scan the image..or keep flicking your eyes hack and forth between each object you identify and the label list.  Imagine how much easier this will be when you can just add a label whilst you're annotating. "ah i forgot this bottle in the corner of the kitchen scene" "ahi forgot this bicycle half hidden behind the bus shelter".  At the moment you have to wait for someone to spot that in "add labels".. which throws things randomly.   (a temporary fix would be a mutual shortcut between unified <->add labels, so once you've found a gap you can fix it, and go back to annotating immediately)

**Recommendations - Other ways for Casual users to contribute**

- whole image classification - scene labels. The difference between 'highstreet', 'residential street', 'town, village' etc.. 

- comparative modes - show 2 images with the same scene labels, and ask which one represents each more (r.e.  initially difficult judgements like 'town vs village'.. these could be % arrived at by votes?)

- thematic judgements.  Seasons. Night/Day

- states .. what is the person doing (sitting, walking,sleeping,standing..)

- refinements: Once a pro user has drawn around an object, the label could be 'walked' through the graph as a quiz

- Bounding boxes. "draw a box around the most interesting objects", or "draw a box around all the heads", "draw a box around as many individual vehicles as you can.."

- Components: once you've got a box around a large enough object, you could ask the casual user to anotate it's parts

- Touch-the-squares:  approximate boundaries for complex tasks like vegetation 

- "labels per  grid cell" if you could divide the image up into a 2x2 or 3x3 grid, it might alleviate the  problem above ..and it gives yes another training signal

tasks might be ok for things which are universal like "sky"

** Complexity of the labelllist**

Interestingly earlier you drew the 'project Trilema' which I'm familiar with ("time / quality / budget" - pick 2), and raised the "dog / canine / hound.." problem

I think the graph idea should let you "have your cake and eat it": you could do verification gradually, i.e.  if you have a "fox" you can first verify that it's a "canine" (in a page showing lots of "canine" examples.. (canine=dog,fox,wolf,jackal..)

I think the graph+refinement should let you have an ever expanding list, such that the resolution of the dataset can expand through time(see wikipedia red-links. a red-link is a "todo")

**Comment by bbernhard** [comment]
After thinking about it again (yeah, sometimes sleep helps a bit), your arguments make more sense to me. You are right, there are some (long term) changes needed in order to re-position the tasks approach. However, as a (more suitable) replacement will probably take a while, I hope we can find a way to improve the unified mode, while not obsoleting the tasks based approach completely (we can rework/remove/restructure the tasks based approach gradually over time).

There are a few changes needed in order to fully support free labeling in combination with annotations,
so this won't work tomorrow, but I think in a few weeks we could be ready.

Apart from the implementation work that needs to be done, there are two open problems, that I would like to address first. In my opinion those are general problems - so not necessarily related to the unified mode/tasks based approach.

# Problem 1

**What do we want so solve?**

differentiate between scene labels/meta labels and normal labels.

**Why do we need to solve that?**

- it pollutes the normal annotation mode with a lot of labels that are not meant to be annotatable

- I think at a certain number it will even become a problem in the unified mode - scene labels will pollute the view and you have to scroll over the labels to find the ones that are meant to be able to annotatable.

**How can we solve that?**

- add own section/list for scene labels/metalabels
- add a keyboard shortcut to mark them somehow? (e.q `/s still image`). As soon as the label is entered, the `/s` will be stripped away. We can now use the information that it is a scene label to display the label differently (maybe collapsed?)


# Problem 2

**What do we want to solve?**

How can we make sure that people are actively using the properties system for refinements?

**Why do we need to solve that?**

You are totally right, in principial there is no need for that, as we can always "label the labels".

From a personal point of view however, I would really like to see things like color information, size of the object, material etc represented as property and not part of the label name.

In my opinion there is not really a reason to put that in the label name. The only argument I can come up with is: it's easier to do, as you don't have to switch back and forth between keyboard and mouse. But if that's the main reason we should definitely adapt the UI to make that easier.

Similar, for parts/components. Personally, I would rather see them fully qualified (i.e `tail/dog` instead of just `tail`) as this is (at least in my opinion) a low hanging fruit that helps enormously with structuring. It requires a bit more typing, but doing that properly might save us from big restructuring efforts.

A fully annotated label also makes it possible to hierarchically group labels (e.q: `window/car` would be a child of `car`) in the labels list.

Or am I missing something here and my suggestions are counter-productive?

**How can we solve that?**

I think it's pretty complicated to enforce that, as we would need to parse the label and interpret it.
So, I am wondering if it's enough for now to add a FAQ with best practices and deal with that later?


What do you think about that? Would that work for you? 

Thanks also for your suggestions on how to improve the situation - really appreciated! I can definitely see myself implementing those, although I have to admit that this probably will take some time, as some of them require some rather big changes.

**Comment by dobkeratops** [comment]
right.. i think in terms of steering rather than abrupt change. 
the add-labels screen is still useful to have - and that's how people will add the scene labels;

what I'd imagine is a shortcut (you've got "Goto..") between the unified mode and the current "Add Labels" screen. That would be workable (we could fix the label holes). 

The problem with the current 'goto' is at the moment you click, the labels you just added aren't yet registered as tasks (as far as i can tell) .. if you can get it to a state where you can add a label, then click "goto->unified.." and see that label ready for annotation, and get back ("ah i found this gap, i need another label..") that'll work

**Comment by dobkeratops** [comment]
>differentiate between scene labels/meta labels and normal labels.

>Why do we need to solve that?

>it pollutes the normal annotation mode with a lot of labels that are not meant to be annotatable

right;
certainly at the minute there's an ambiguity and some labels might be both. (fora focussed picture of a dog: "dog" is a valid scene label..  I thought of 'kitchen' as a scene label but i've got a living room with a window into the kitchen, you could annotate the window as 'kitchen' :) . and so on)

as for "pollution", if clicking on the scene labels gave label dynamic recommendations (actually clicking them added to the palette) they'd serve a purpose in the main list. however I do like your idea of seperate lists. (if "dog" is the scene label, the annotatable labels might be dog parts..?)

lets see how things go.. personally I think that label list is going to be fine up until about 16? I haven't found it to be a problem *yet*. you might have 4 un-annotateables (general themes) and 12 real labels. you might find this 'co-occuring label recomendations' idea is useful for any selection

**Comment by bbernhard** [comment]
> The problem with the current 'goto' is at the moment you click, the labels you just added aren't yet registered as tasks (as far as i can tell) .. if you can get it to a state where you can add a label, then click "goto->unified.." and see that label ready for annotation, and get back ("ah i found this gap, i need another label..") that'll work

totally agreed. :)

> certainly at the minute there's an ambiguity and some labels might be both. (fora focussed picture of a dog: "dog" is a valid scene label.. I thought of 'kitchen' as a scene label but i've got a living room with a window into the kitchen, you could annotate the window as 'kitchen' :) . and so on)

> as for "pollution", if clicking on the scene labels gave label dynamic recommendations (actually clicking them added to the palette) they'd serve a purpose in the main list. however I do like your idea of seperate lists. (if "dog" is the scene label, the annotatable labels might be dog parts..?)

> lets see how things go.. personally I think that label list is going to be fine up until about 16? I haven't found it to be a problem yet. you might have 4 un-annotateables (general themes) and 12 real labels. you might find this 'co-occuring label recomendations' idea is useful for any selection

My gut feeling is, that this will become a "problem" pretty fast. At the moment, I think it's less of a problem, as we do not have that much "non-annotatable" labels unlocked. I've checked in the database and there are already quite a few images that have > 16 labels including a few non-annotatable ones. 

Not sure if its the best solution, but I think two listviews might solve the problem. If the label will be added to the metalabels list, it will be non-annotatable. If you later decide that you want to annotate it, just drag-and-drop the label to the first listview. 

Regarding the components/parts: I was hoping that we can use the fully-qualified name (e.g `tail/dog`) for hierarchical structuring/grouping. So, that we end up with a view similar to this one: 

![collapsing](https://user-images.githubusercontent.com/5710631/57974824-b3468080-79be-11e9-8fba-8e5db5343b00.png)

(I think we should limit the number of hierarchy levels however, otherwise this could become unmanageable. I guess in most cases one hierarchy level (parent -> child) should be enough; everything else can be expressed with the label graph)

**Comment by dobkeratops** [comment]
I do like the tree view idea. I suppose you could put the 'un-annotateables' in their own tree node. perhaps you could extend it to use the graph aswell

```
+scene
 +urban
  +city street
+ground
 +road
 +pavement
 +traffic island
 +central reservatoin
 +driveway
 +cyclepath
+road_vehicle
 +car
  +hatchback car
  +saloon car
 +truck
  +tanker truck
  +cement mixer truck
+vegetation
 +grass
 +tree
 +bush
+person
 +pedestrian
 +cyclist
 +driver
```
> it pollutes the normal annotation mode with a lot of labels that are not meant to be annotatable

personally I dont think this *won't* be a problem,

..because in a truly unified mode - you're going to type in the name of a label and annotate with it immediately.. or the other way round- draw around something, then type in what it is (easy shortcut to re-use the last label).

by growing it unified, the list will be sorted from most obvious to most obscure. (you'll do the most obvious, prominent object first)

As it stands I'm seeing label lists ~20 in the most complex current cases; even thatll be doable with a simple scrolling list.. I used paint programs with 16-32 color palettes no problem (320x240 screen) but bear in mind the above context - most of thetime you're working, that list will be shorter (because you haven't entered the labels yet).

you'll always be able to choose a label by typing it in (if it happened to already be in the list.. great) ("add label and make current..") .. and that'll be the default way

**Comment by bbernhard** [comment]
> personally I dont think this will be a problem,

> ..because in a truly unified mode - you'll be able to type in the name of a label and annotate with it immediately.. or the other way round- draw around something, then type in what it is (easy shortcut to re-use the last label).

yeah, right. But I was hoping that we can maybe let the task based approach and the unified mode live in (happy) co-existence (at least for some time). That way, both user groups could add labels that the other one is then working on (e.g: someone adds a label in the unified view, but doesn't want to annotate it, so maybe someone from the "task based" user group will take the task. Similar the other way round: One adds a label in the label view, but doesn't want to annotate it - it could then be done in the unified mode). 

I think we are almost there concept-wise..the only thing that would be needed is to mark the label as "metalabel/scene label". 

Apart from that, we probably should also think about highlighting the labels, that are missing an annotation (at the moment you have to click on every label in the unified mode label list to see if it is already annotated; that's of course not a problem if you add the label and annotate it immediately, but for labels that are already existing but missing an annotation, it could be interesting).

Regarding the tree view: Personally, I would really like to keep the labels list as flat as possible (max. 1-2 hierarchy levels). In my opinion the best is, to just add enough information that the label will become unique. e.g: `leg` is bad, as it could be `leg/person`, ` leg/dog`, `leg/cat`, ...etc. But if you add a label `leg/person` it already contains all the information that's needed to position it in the label graph.

Of course, one could also add the following label `leg/dog/animal` (or even `leg/dog/animal/mammal`), but personally I feel that's already too much. The info that `dog` is an animal should be part of the label graph - moving the info down to the image level would make the label re-structuring way more complicated. 

At the moment we have a nicely layered approach. We have the labels (which are pretty flat) and on top of that, we have the label graphs that are user-driven (so, different users could create different label graphs to model different use cases). So, if we move the hierarchical structuring down to the image level, we break the label graph concept. 

Does that make sense?

**Comment by dobkeratops** [comment]
Mock-up .. ideas for how you could manage and visualise large numbers of labels. I think this sort of thing would take some effort to write and refine, but I would just like to illustrate why I think it’s a solveable problem. I imagine togglesble outlines, and perhaps an optimum guess at color  coding for the more widespread lablels.
![IMG_1199](https://user-images.githubusercontent.com/1120754/58067560-9c9e5600-7b85-11e9-91c6-c2bcdf15b2d2.jpeg)

**Comment by dobkeratops** [comment]
>From a personal point of view however, I would really like to see things like color >information, size of the object, material etc represented as property and not part of the label name.

>In my opinion there is not really a reason to put that in the label name. The only argument >I can come up with is: it's easier to do, as you don't have to switch back and forth >between keyboard and mouse. But 

Could the Property system make it easier to auto parse some obvious prefixes.. and couldn’t we have aliases for the common cases.. the thing is at the moment we’re waiting for that UI to refine and aliases just added to the list could enable adding all that informstion right now; I was for example adding variations on sky eg night sky, overcast sky which give informstion about the overall condition. I was also going to suggest that a couple of really common objects = box,cylinder, bin,crate,,container even could benefit from some common material prefixes (wooden box, metal box,plastic box). It’s so easy to just say that in one breath rather than teach a new UI.

**Comment by bbernhard** [comment]
> Could the Property system make it easier to auto parse some obvious prefixes.. and couldn’t we have aliases for the common cases.. the thing is at the moment we’re waiting for that UI to refine and aliases just added to the list could enable adding all that informstion right now; I was for example adding variations on sky eg night sky, overcast sky which give informstion about the overall condition. I was also going to suggest that a couple of really common objects = box,cylinder, bin,crate,,container even could benefit from some common material prefixes (wooden box, metal box,plastic box). It’s so easy to just say that in one breath rather than teach a new UI.

So, you mean you enter a label and that will automatically be broken down into label + property? 

That would be really cool..although I am not sure how the workflow would then look like in case there are multiple objects with different properties. e.g: imagine three boxes with different materials. At the moment, you would draw three bounding boxes and assign each box a different material property. 

Now, lets suppose we have a parser that can separate labels and properties. How would the workflow then look like with the unified mode? Would you then have to enter the labels three times? add label `wooden box` -> annotate -> add label `metal box` -> annotate -> add label `plastic box` -> annotate.

I think it's also a question which UI you are using. In case you are adding the labels in the labels view, I think it could be easier to do it like that: `wooden box`, `metal box`, `plastic box`

In the unified mode however, I personally would prefer to see only the base label `box` and represent everything else with a property. In my opinion the separation between label + property makes it a lot easier to later enrich annotations with additional information. (e.g add color information, information about size, foreground/background info, ..etc). 

What we probably could do as a intermediate solution: Keep the fine granular label (`wooden box`) in the labels view, but rename the label to the base label (`box`) when the label gets unlocked. We do not have the necessary code in place yet to do so, but I think from a technical point of view it should work (at least in a semi-automated way)

That way you could continue adding labels. But that of course only works as long as we manually unlock labels (in that case we can rename the labels before unlocking). As soon as we remove that, we should have figured out, how we deal with that. 

I think it definitely makes sense to have both label granularities (detailed labels in the label view, that describe the object + base label & properties in the annotation view). But not sure if we can have both, without doing the work twice.

**Comment by dobkeratops** [comment]
> gh I am not sure how the workflow would then look like in case there are multiple objects with different properties. e.g: imagine three boxes with different materials. 

What I would imagine there is you’d make a label for each. If your system needs those to be grouped into box with 3 material instances.. is that doable each time you enter a label? (Presumably you’d have to be able to add time meterials separately).. or would you store the material with the polygon?

I guess the msterial/property system is going to come into its own when there’s many eg persons viewpoint, action, ethnicity,clothing,gender.. that’s probably where the dedicated system and UI will have maximum utility. That would indeed get hard to parse as there’d be many ways of saying it in natural language

**Comment by bbernhard** [comment]
> What I would imagine there is you’d make a label for each. If your system needs those to be grouped into box with 3 material instances.. is that doable each time you enter a label? (Presumably you’d have to be able to add time meterials separately).. or would you store the material with the polygon?

At the moment, all the properties are tied to one specific polygon. So you can have three different `apple` polygons in one image and every polygon has a different set of properties. But that's just the internal representation in the database...in case we want to expose it differently in the UI, we can definitely do that.  


> I guess the msterial/property system is going to come into its own when there’s many eg persons viewpoint, action, ethnicity,clothing,gender.. that’s probably where the dedicated system and UI will have maximum utility. That would indeed get hard to parse as there’d be many ways of saying it in natural language

yeah, right - that's exactly what I would like to see the properties system become.

Personally, I was hoping that we can get the unified mode + properties in a state where we have everything (or at least everything we can think of) figured out (at least concept-wise), before we enable the free labeling + annotation workflow. I think we aren't that far away from that goal anymore... 

If we accomplish that, I think we've solved a bunch of problems in one go and created a _really_ powerful system.

**Comment by bbernhard** [comment]
just a short update: I am currently extending the unified mode with the ability to add labels directly in there. In a first iteration, only unlocked (i.e productive) labels will be allowed (that's because there are some backend changes needed to support free annotation, which need to be implemented first). In a next iteration free annotation will follow. :)

**Comment by dobkeratops** [comment]
That’s awesome.. even without free labelling, being able to add productive labels will be a huge step forward.. greatly looking forward to it. At that point the software will have crossed a major threshold... getting a larger productive-label list is “just” data, (ok data can still be difficult,, but there’s other resources to draw on like other datasets, wordnet, etc. I’m confident we can figure out a more comprehensive label list eventually )

**Comment by bbernhard** [comment]
another short update: 

I think I am almost done with the changes for the first iteration. I still need to test the thing a bit more, but hopefully I can push that to production next weekend. 

Here's a preview (please ignore the nonsense labels, they are just for testing ;))

![Selection_007](https://user-images.githubusercontent.com/5710631/58760695-d4b97780-853b-11e9-91a4-b0fa54871d05.png)

**Comment by dobkeratops** [comment]
That looks awesome. I note you can also reach the unified view through “annotate->expert mode..”, that’s useful.

I wonder if in future you could eventually present tasks and labelling modes as simplifications of the unified view.. eg imagine if the label list just happened to start hidden (its still firing up with an initial task..), and vica versa what if the annotation tools were just hidden  in the “add labels view..”.  .... not worth worrying about now.. just an idea for future refinement - and of course the dedicated views might still have more room for other things (label suggestion lists, of the image taking up more space in the annotation mode)

Another little idea would reference images to the side .. imagine if once you have selected a scene label it would show an example with label ideas (suggestions by comparsion/example)... those could be smaller images to the side as you wouldn’t need to work on them

**Comment by bbernhard** [comment]
> I wonder if in future you could eventually present tasks and labelling modes as simplifications of the unified view.. eg imagine if the label list just happened to start hidden (its still firing up with an initial task..), and vica versa what if the annotation tools were just hidden in the “add labels view..”. .... not worth worrying about now.. just an idea for future refinement - and of course the dedicated views might still have more room for other things (label suggestion lists, of the image taking up more space in the annotation mode)

I  have to admit, that I am not really happy with the annotation dropdown menu - the separation between "simple mode" and "expert mode" makes the dropdown menu look really crowded. That's why I've removed the "expert mode" functionality from the annotation menu and made it a top level entry. 

In my dev version, it looks like this: 

![Selection_008](https://user-images.githubusercontent.com/5710631/58766075-c8a0da80-857a-11e9-8780-ba5ecf5e3b43.png)

Any suggestions regarding the name are really welcome ;)

I think that aligns pretty much with what you had in mind? :)

> Another little idea would reference images to the side .. imagine if once you have selected a scene label it would show an example with label ideas (suggestions by comparsion/example)... those could be smaller images to the side as you wouldn’t need to work on them

like that! 

Although I've to admit that I am not sure about the exact placement of those images. Ideally we find position where

-) the reference images doesn't distract the user
-) most of the workspace space is preserved
